source/ecs-image-handler/src/processor/image/crop.ts

```diff
@@ -0,0 +1,136 @@
+import { Metadata } from 'sharp';
+import { IImageAction, IImageContext } from '.';
+import { IActionOpts, ReadOnly, InvalidArgument } from '..';
+
+export interface CropOpts extends IActionOpts {
+  w: number;
+  h: number;
+  x: number;
+  y: number;
+  g: string;
+}
+
+export class CropAction implements IImageAction {
+  public readonly name: string = 'crop';
+
+  public validate(params: string[]): ReadOnly<CropOpts> {
+    var opt: CropOpts = { w: 0, h: 0, x: 0, y: 0, g: 'nw' };
+
+    if (params.length < 2) {
+      throw new InvalidArgument('Crop param error, e.g: crop,x_100,y_50');
+    }
+
+    for (const param of params) {
+      if ((this.name === param) || (!param)) {
+        continue;
+      }
+      const [k, v] = param.split('_');
+      if (k === 'w') {
+        const w = parseInt(v);
+        if (w > 0) {
+          opt.w = w;
+        } else {
+          throw new InvalidArgument('Crop param \'w\' must be greater than  0');
+        }
+      } else if (k === 'h') {
+        const h = parseInt(v);
+        if (h > 0) {
+          opt.h = h;
+        } else {
+          throw new InvalidArgument('Crop param \'h\' must be greater than  0');
+        }
+      } else if (k === 'x') {
+        const x = parseInt(v);
+        if (x >= 0) {
+          opt.x = x;
+        } else {
+          throw new InvalidArgument('Crop param \'x\' must be greater than or equal to 0');
+        }
+      } else if (k === 'y') {
+        const y = parseInt(v);
+        if (y >= 0) {
+          opt.y = y;
+        } else {
+          throw new InvalidArgument('Crop param \'y\' must be greater than or equal to 0');
+        }
+      } else if (k === 'g') {
+        if (v == 'nw' || v == 'north' || v == 'ne' ||
+          v == 'west' || v == 'center' || v == 'east' ||
+          v == 'sw' || v == 'south' || v == 'se') {
+          opt.g = v;
+        } else {
+          throw new InvalidArgument('Crop param \'g\' must be  \'nw, north, ne, west, center, east, sw, south, se\'');
+        }
+      } else {
+        throw new InvalidArgument(`Unkown param: "${k}"`);
+      }
+
+    }
+    return opt;
+  }
+
+
+  public async process(ctx: IImageContext, params: string[]): Promise<void> {
+    const opt = this.validate(params);
+
+    var height = opt.h;
+    var width = opt.w;
+    var x = opt.x;
+    var y = opt.y;
+
+
+    await ctx.image.metadata()
+      .then(function (metadata: Metadata) {
+
+        if (metadata.height == undefined || metadata.width == undefined) {
+          throw new InvalidArgument('Incorrect image format');
+        }
```

Comment:
No need to use `then` since we have `async/await`

```ts
const metadata = await ctx.image.metadata();
if (metadata.height == undefined || metadata.width == undefined) {
    throw new InvalidArgument('Incorrect image format');
}
// ...
```